testbed/tests/app/test_desktop.py

```diff
@@ -370,25 +368,45 @@ async def test_current_window(app, app_probe, main_window):
     window2.content = toga.Box(style=Pack(background_color=CORNFLOWERBLUE))
     window3.content = toga.Box(style=Pack(background_color=FIREBRICK))
 
-    # We don't need to probe anything window specific; we just need
-    # a window probe to enforce appropriate delays.
-    window1_probe = window_probe(app, window1)
-
     window1.show()
     window2.show()
     window3.show()
 
-    await window1_probe.wait_for_window("Extra windows added")
+    await main_window_probe.wait_for_window("Extra windows added")
+
+    # When a window without any dialog is made the current_window,
+    # then `app.current_window` should return the specified window.
+    app.current_window = window1
+    await main_window_probe.wait_for_window("Window 1 is current")
+    if app_probe.supports_current_window_assignment:
+        assert app.current_window == window1
+
+    # When a dialog is in focus, `app.current_window` should
+    # return the window from which the dialog was initiated.
+
+    window2_info_dialog = toga.InfoDialog("Second Window Info", "Some info")
+    app_probe.setup_info_dialog_result(window2_info_dialog)
 
     app.current_window = window2
-    await window1_probe.wait_for_window("Window 2 is current")
+    window2_dialog_task = app.loop.create_task(window2.dialog(window2_info_dialog))
+    await main_window_probe.wait_for_window("Window 2 is current")
     if app_probe.supports_current_window_assignment:
         assert app.current_window == window2
+    await app_probe.redraw("select 'OK'")
+    # Cancel the task to avoid dangling
+    window2_dialog_task.cancel()
+
+    window3_info_dialog = toga.InfoDialog("Third Window Info", "Some info")
+    app_probe.setup_info_dialog_result(window3_info_dialog)
 
     app.current_window = window3
-    await window1_probe.wait_for_window("Window 3 is current")
+    window3_dialog_task = app.loop.create_task(window3.dialog(window3_info_dialog))
+    await main_window_probe.wait_for_window("Window 3 is current")
     if app_probe.supports_current_window_assignment:
         assert app.current_window == window3
+    await app_probe.redraw("select 'OK'")
+    # Cancel the task to avoid dangling
+    window3_dialog_task.cancel()
```

Comment:
I don't think we need to do the "in the presence of a dialog" test multiple times. We should still be testing the "normal" case of "just assigning a new window"; we only need to do one additional check of "...and if there's a current dialog, that doesn't get in the way".

We *should*, however, confirm that when a dialog is dismissed, but the current window *hasn't* been modified that the current window is still what it should be.

Reply:
I agree, I have made the "in presence of dialog test" to be tested only on window1.